content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,260 @@
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+    * When requests.memory is set, limits.memory is not set, we substitute limits.memory
+    for node allocatable memory in the formula:
+    ![](./container-memory-high-no-limits.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+3. BestEffort The pod gets a BestEffort class if limits.memory and requests.memory are not 
+set. We set requests.memory = 0 and substitute limits.memory for node allocatable memory 
+in the formula:
+
+    ![](./container-memory-high-best-effort.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+### Cgroup Hierarchy
+When container memory requests are made, kubelet passes `memory.min` to the back-end 
+CRI runtime (possibly containerd, cri-o) via the `Unified` field in CRI during 
+container creation. The `memory.min` in container level cgroup will be set to:  
+
+![](./container-memory-min.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+Since the `memory.min` interface requires that the ancestor cgroup directories are all 
+set, the pod and node cgroup directories need to be set correctly. 
+
+`memory.min` in pod level cgroup:  
+![](./pod-memory-min.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+`memory.min` in node level cgroup:  
+![](./node-memory-min.svg)  
+<sub>i: the i<sup>th</sup> pod in one node, j: the j<sup>th</sup> container in 
+one pod</sub>
+
+Kubelet will manage the cgroup hierarchy of the pod level and node level cgroups 
+directly using runc libcontainer library, while container cgroup limits are managed by the container runtime.
+
+## How do I use it?
+Here are the prerequisites for enabling Memory QoS on your Linux node, some of these are 
+related to [Kubernetes support for cgroup v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2).
+
+1. Kubernetes since v1.27
+2. [runc](https://github.com/opencontainers/runc) since v1.0.0-rc93; [containerd](https://containerd.io/)
+since 1.4; [cri-o](https://cri-o.io/) since 1.20
+3. Linux kernel minimum version: 4.15, recommended version: 5.2+
+4. Linux image with cgroupv2 enabled or enabling cgroupv2 unified_cgroup_hierarchy manually
+
+OCI runtimes such as runc and crun already support cgroups v2 [`Unified`](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#unified), and Kubernetes CRI has also made the desired changes to support passing [`Unified`](https://github.com/kubernetes/kubernetes/pull/102578). However, CRI Runtime support is required as well. Memory QoS in Alpha phase is designed to support containerd and cri-o. Related PR [Feature: containerd-cri support LinuxContainerResources.Unified #5627](https://github.com/containerd/containerd/pull/5627) has been merged and released in containerd 1.6. CRI-O [implement kube alpha features for 1.22 #5207](https://github.com/cri-o/cri-o/pull/5207) has been merged and released in 1.22.
```

Comment:
```suggestion
OCI runtimes such as crun and runc already support cgroups v2 [`Unified`](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#unified),
and Kubernetes has also made the desired CRI changes to support passing
[`Unified`](https://github.com/kubernetes/kubernetes/pull/102578).
However, support is required in CRI-compatible runtimes as well.
Memory QoS in alpha phase is designed to support containerd and cri-o.
A related PR, [Feature: containerd-cri support LinuxContainerResources.Unified](https://github.com/containerd/containerd/pull/5627), has been merged and released in containerd 1.6
 CRI-O PR [implement kube alpha features for 1.22](https://github.com/cri-o/cri-o/pull/5207) has been
 merged and released in CRI-O 1.22.
```
Several things going on here. We list 3rd party products in alphabetical order, for example.